### PR TITLE
Slither updates

### PR DESCRIPTION
**Pull request by area**
Fixes issues the newest version of slither raises. This includes reentrancy risks inside loops, variables that could be immutable but weren't, and what I think I agree was an unnecessary external call to `version` in the `upgrade` function. Not convinced I've fixed these in the best possible ways, so open to suggestions.

### CODE REVIEW

File: contracts/colony/Colony.sol [hunk 1]
```diff
@@ -109,8 +109,11 @@ contract Colony is BasicMetaTransaction, Multicall, ColonyStorage, PatriciaTreeP
       require(uint256(_amounts[i]) <= fundingPots[1].balance[token], "colony-bootstrap-not-enough-tokens");
       fundingPots[1].balance[token] = sub(fundingPots[1].balance[token], uint256(_amounts[i]));
       nonRewardPotsTotal[token] = sub(nonRewardPotsTotal[token], uint256(_amounts[i]));
+    }
 
-      assert(ERC20Extended(token).transfer(_users[i], uint256(_amounts[i])));
+    // After doing all the local storage changes, then do all the external calls
+    for (uint i = 0; i < _users.length; i++) {
```
Comment (kronosapiens) [inline contracts/colony/Colony.sol:115]: Is `i` being declared twice here? Stylistically I would use `j` to indicate a second loop but that's just me.

File: contracts/colony/ColonyAuthority.sol [hunk 2]
```diff
@@ -28,6 +28,8 @@ contract ColonyAuthority is CommonAuthority {
   uint8 constant ARCHITECTURE_ROLE = uint8(ColonyDataTypes.ColonyRole.Architecture);
   uint8 constant ROOT_ROLE = uint8(ColonyDataTypes.ColonyRole.Root);
 
+  // colony is used in the constructor by calls to addRoleCapability, despite what slither thinks
```
Comment (kronosapiens) [inline contracts/colony/ColonyAuthority.sol:31]: I don't see where...

**Review by kronosapiens — COMMENTED** [review]
Looks reasonable to me, will take another look on Monday

File: contracts/colony/Colony.sol (hunk 1 above)
Comment (kronosapiens) [reply contracts/colony/Colony.sol:115]: Oh, can you declare the same identifier twice in different loops? I assumed they were in the same scope, guess not.

Comment (area) [reply contracts/colony/Colony.sol:115]: Yep, they are [different scopes](https://docs.soliditylang.org/en/v0.8.17/control-structures.html#scoping-and-declarations):

> As an exception to this rule, variables declared in the initialization part of a for-loop are only visible until the end of the for-loop.

File: contracts/colony/ColonyAuthority.sol (hunk 2 above)
Comment (area) [reply contracts/colony/ColonyAuthority.sol:31]: https://github.com/JoinColony/colonyNetwork/blob/b7201a0b0051ed9a985a0a15e73779b9ced5d780/contracts/colony/ColonyAuthority.sol#L136 - this call in `addRoleCapability` reads the `colony` storage variable.

File: contracts/colony/Colony.sol (hunk 1 above)
Comment (kronosapiens) [inline contracts/colony/Colony.sol:115]: Can we make it `uint256` for consistency with the rest of the codebase?

File: contracts/colony/ColonyRoles.sol [hunk 3]
```diff
@@ -104,20 +104,25 @@ contract ColonyRoles is ColonyStorage, ContractRecoveryDataTypes {
...
-          emit RecoveryRoleSet(_user, setTo);
-        }
         emit ColonyRoleSet(msgSender(), _user, _domainId, roleId, setTo);
 
       }
       roles >>= 1;
       rolesChanged >>= 1;
     }
+
+    // Update the storage slot tracking number of recovery roles after all the external calls are complete
```
Comment (kronosapiens) [inline contracts/colony/ColonyRoles.sol:114]: Shouldn't this be before the external calls?

Comment (area) [reply contracts/colony/ColonyRoles.sol:114]: Yup, I really messed up there :disappointed:

File: contracts/colony/ColonyAuthority.sol (hunk 2 above)
Comment (area) [reply contracts/colony/ColonyAuthority.sol:31]: Prior to [solidity 0.8.8](https://github.com/ethereum/solidity/releases/tag/v0.8.8) immutable variables could not be read in the constructor. One to be picked up in #1053

Comment (kronosapiens) [reply contracts/colony/ColonyAuthority.sol:31]: Ah, fascinating.

**Review by kronosapiens — APPROVED** [review]
Fab

**Pull request merged**